source/common/logger/ur_logger.hpp

```diff
@@ -0,0 +1,122 @@
+// Copyright (C) 2023 Intel Corporation
+// SPDX-License-Identifier: MIT
+
+#ifndef UR_LOGGER_HPP
+#define UR_LOGGER_HPP 1
+
+#include <algorithm>
+#include <memory>
+
+#include "ur_level.hpp"
+#include "ur_sinks.hpp"
+#include "ur_util.hpp"
+
+namespace logger {
+
+class Logger {
+  public:
+    Logger(std::unique_ptr<logger::Sink> sink) : sink(std::move(sink)) {
+        if (!this->sink) {
+            throw std::invalid_argument(
+                "Can't create logger with nullptr Sink!");
+        }
+        this->level = logger::Level::QUIET;
+    }
+
+    Logger(logger::Level level, std::unique_ptr<logger::Sink> sink)
+        : level(level), sink(std::move(sink)) {
+        if (!this->sink) {
+            throw std::invalid_argument(
+                "Can't create logger with nullptr Sink!");
+        }
+    }
+
+    ~Logger() = default;
+
+    void setLevel(logger::Level level) { this->level = level; }
+
+    template <typename... Args> void debug(const char *format, Args &&...args) {
+        log(logger::Level::DEBUG, format, std::forward<Args>(args)...);
+    }
+
+    template <typename... Args> void info(const char *format, Args &&...args) {
+        log(logger::Level::INFO, format, std::forward<Args>(args)...);
+    }
+
+    template <typename... Args>
+    void warning(const char *format, Args &&...args) {
+        log(logger::Level::WARN, format, std::forward<Args>(args)...);
+    }
+
+    template <typename... Args> void error(const char *format, Args &&...args) {
+        log(logger::Level::ERR, format, std::forward<Args>(args)...);
+    }
+
+  private:
+    logger::Level level;
+    std::unique_ptr<logger::Sink> sink;
+
+    template <typename... Args>
+    void log(logger::Level level, const char *format, Args &&...args) {
+        if (level < this->level) {
+            return;
+        }
+
+        sink->log(logger::level_to_str(level), format,
+                  std::forward<Args>(args)...);
+    }
+};
+
+std::unique_ptr<Logger> logger;
```

Comment:
Will this work correctly? I think you should use either static (if you want this to be per compilation unit) or extern.
But do we actually need to have this as a global variable? Perhaps we could store the logger in contexts? (e.g. `loader::context` and `ur_lib::context`).

Reply:
Right, for common, we still need a global variable.

As for optional - it would work as well, but it requires C++17 (and we have C++14 I think).

Reply:
C++14 I think comes from being compiled as part of LLVM. [C++17 is now the default in LLVM](https://www.phoronix.com/news/LLVM-Clang-16-Default-GNU17) so, assuming https://github.com/intel/llvm has those patches, I'd be open to changing to C++17.